common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java

```diff
@@ -1509,12 +1509,66 @@ public boolean semanticEquals(final UTF8String other, int collationId) {
     return CollationFactory.fetchCollation(collationId).equalsFunction.apply(this, other);
   }
 
+  private interface SubstringEquals {
+    boolean equals(UTF8String left, UTF8String right, int posLeft, int posRight,
+      int lenLeft, int lenRight);
+  }
+
+  private static class ByteSubstringEquals implements SubstringEquals {
+    @Override
+    public boolean equals(UTF8String left, UTF8String right, int posLeft, int posRight,
+      int lenLeft, int lenRight) {
+      if (lenLeft != lenRight || left.getByte(posLeft) != right.getByte(posRight)) {
+        return false;
+      }
+      return (ByteArrayMethods.arrayEquals(left.base, left.offset + posLeft, right.base,
+        right.offset + posRight, lenLeft));
+    }
+  }
+
+  private static final ByteSubstringEquals byteSubstringEquals = new ByteSubstringEquals();
+
+  private static class CollationSubstringEquals implements SubstringEquals {
```

Comment:
have you actually tested the impact of having a smaller number of allocations? Because creating a UTF8 string might not be that expensive, especially compared to converting from UTF8String to String which will happen anyway

Reply:
It's true that `UTF8String` is a small object (base, offset and length), but the issue with Levenshtein distance is that the algorithm has quadratic complexity. So when comparing two strings of length 100, you would allocate 10k objects which is a waste.